### PR TITLE
perf: dynamically batch tx sender recovery

### PR DESCRIPTION
**Pull request by onbjerg**
The performance regression in the sender recovery stage was caused by us effectively queuing 5000 really "fast" (relatively) jobs, leading to a lot of time lost on Rayon's worker threads trying to steal more jobs.

The solution is to reintroduce batching. For now, we create batches based on the number of worker threads in the Rayon threadpool. This works since we are limited by memory, and can't crank the commit threshold too much, and separate config for batch sizes in this case doesn't make much sense.

This is a perf grab of the current sender recovery stage:

Screenshot from 2023-03-18 13-19-39

As we can see here (on the top right), almost 50%(!) of the time is spent trying to get more work.

Compare this with this PR:

Screenshot from 2023-03-18 14-01-03

We almost spend no time trying to get more work.

The speedup for me is that sender recovery now feels snappy again - before, it felt like it took 20-30s per 5k blocks, now it takes about 3-4.

### CODE REVIEW

File: crates/stages/src/stages/sender_recovery.rs [hunk 1]
```diff
@@ -75,36 +76,49 @@ impl<DB: Database> Stage<DB> for SenderRecoveryStage {
...
+                        let (tx_id, transaction) = entry?;
+                        let sender = transaction.recover_signer().ok_or(StageError::from(
+                            SenderRecoveryStageError::SenderRecovery { tx: tx_id },
+                        ))?;
+
+                        Ok((tx_id, sender))
+                    })
+                    .for_each(|result: Result<_, StageError>| {
+                        let _ = tx.send(result);
+                    });
```
Comment (mattsse) [inline crates/stages/src/stages/sender_recovery.rs:115]: sending them one by one is fine we

File: crates/stages/src/stages/sender_recovery.rs [hunk 2]
```diff
@@ -75,36 +76,49 @@ impl<DB: Database> Stage<DB> for SenderRecoveryStage {
...
-        for entry in entries {
-            let (tx_id, transaction) = entry?;
+        // Spawn recovery jobs onto the default rayon threadpool and send the result through the
+        // channel.
+        //
+        // We try to evenly divide the transactions to recover across all threads in the threadpool.
+        // Chunks are submitted instead of individual transactions to reduce the overhead of work
+        // stealing in the threadpool workers.
+        for chunk in
+            &tx_walker.chunks(self.commit_threshold as usize / rayon::current_num_threads())
```
Comment (mattsse) [inline crates/stages/src/stages/sender_recovery.rs:94]: 👍

in hindsight this is kinda obvious

**Review by mattsse — APPROVED** [review]
gg

File: crates/stages/src/stages/sender_recovery.rs (hunk 2 above)
Comment (onbjerg) [reply crates/stages/src/stages/sender_recovery.rs:94]: yeah...

**Pull request merged**